nmea_parser.c

```diff
@@ -1,191 +1,315 @@
-						val = (char *) malloc(strlen(ptr));
-						strncpy(val,ptr,strlen(ptr));
-						setDegradation(atof(val));
-						debug_print("Get Bugs Value: %f\n",atof(val));
-					}
-				break;
-				
-				case 'P':
-					//Set Polar
-					// depreciated
-					if (*(ptr+1) == 'O' &&  *(ptr+2) == 'L') {
-						ptr = strtok(NULL, delimiter);
-						val = (char *) malloc(strlen(ptr));
-						strncpy(val,ptr,strlen(ptr));
-						polar.a=atof(val);
-						ptr = strtok(NULL, delimiter);
-						strncpy(val,ptr,strlen(ptr));
-						polar.b=atof(val);
-						ptr = strtok(NULL, delimiter);
-						strncpy(val,ptr,strlen(ptr));
-						polar.c=atof(val);
-						ptr = strtok(NULL, delimiter);
-						strncpy(val,ptr,strlen(ptr));
-						polar.w=atof(val);
-						setPolar(polar.a, polar.b, polar.c, polar.w);
-					}
-				break;
 
-				case 'R':
-					//Set Real Polar
-					if (*(ptr+1) == 'P' &&  *(ptr+2) == 'O') {
-						if (read_float_from_sentence(3,fvals,NULL,delimiter)) {
-							// convert horizontal speed from m/s to km/h
-							polar.a=fvals[0]/(3.6*3.6);
-							polar.b=fvals[1]/3.6;
-							polar.c=fvals[2];
-							setRealPolar(polar.a, polar.b, polar.c);
-							debug_print("Get Polar RPO: %f, %f, %f\n",
-										polar.a,polar.b,polar.c);
-						}
-					}
-				break;
-
-				case 'S':
-					if (*(ptr+1) == 'T' && *(ptr+2) == 'F') {
-						// Set STF Mode
-						debug_print("Set STF Mode\n");
-						set_vario_mode(stf);
-					}
-				break;
-				case 'V':
-					if (*(ptr+1) == 'U') {
-						// volume up
-						debug_print("Volume up\n");
-						change_volume(+10.0);
-					}
-					if (*(ptr+1) == 'D') {
-						// volume down
-						debug_print("Volume down\n");
-						change_volume(-10.0);
-					}
-					if (*(ptr+1) == 'M') {
-						// Toggle Mute
-						debug_print("Toggle Mute\n");
-						toggle_mute();
-					}
-					if (*(ptr+1) == 'A' &&  *(ptr+2) == 'R') {
-						// Set Vario Mode
-						debug_print("Set Vario Mode\n");
-						set_vario_mode(vario);
+	buffer[2000]=0;
+	if ((ptr2=strchr(buffer,'$'))!=NULL) {
+		for (i=0;ptr2[i]!=0;++i) buffer[i]=ptr2[i];
+		buffer[i]=0;
+	}
+	if (strlen(buffer)+strlen(message)>1999) 
+		strcpy(buffer,message);
+	else strcpy(buffer+strlen(buffer),message); // Append the new sentence to the last
+	ptr = strtok(buffer, del1); // Tokenize about '$'
+	last=buffer;
+	while (ptr != NULL) {
+		last=ptr;
+		tick=1;
+		if ((*ptr=='P') && (*(ptr+1)=='O') && (*(ptr+2)=='V') && (*(ptr+3)==',')) { // If you begin the sentence correctly  
+			if (verify_checksum(ptr)) { // And checksum passes
+				tick=0;
+				ptr2=ptr+4; // Start looking after the $POV
+				ptr+=strlen(ptr)+1; // Move ptr to the next POV
+				ptr2=strtok(ptr2,del2); // Tokenize about ','
+				while (ptr2 != NULL) { // Repeat until you hit the next '$'  
+					switch (*ptr2) {
+						case 'C':
+							// Command sentence
+							// get next value
+							ptr2 = strtok(NULL, del2);
+							switch (*ptr2) {
+								case 'M':
+									if (*(ptr2+1) == 'C') {
```

Comment:
How do you know ptr2+1 is pointing to a valid position in the sentence? Where is the end of the sentence?

Reply:
The obvious answer is, you are only looking at this after establishing that the sentence is terminated by a valid checksum.  That means there have to be at least three characters after a comma so this therefore *ptr2+1, *ptr2+2, *ptr2+3 by definition have to be valid.

Reply:
A valid checksum doesn't say much about the length of a sentence.

Reply:
ptr2=strtok(NULL,del2) will search for a comma.  By definition a checksum is a '*' followed by two hexadecimal characters at the END of the sentence.  That means any comma in the sentence must be proceeded by at least 3 characters prior to a 0 terminator.  So there will ALWAYS be at least 4 valid characters (including the 0 terminator) after ptr2 is set in line 160.

Reply:
You're missing the point:
What about the values after these characters? How do you know the sentence is long enough?